File: 0629-k-inverse-pairs-array/0629-k-inverse-pairs-array.java
class Solution {
    int M = 1000000007;

    public int kInversePairs(int n, int k) {
        int[][] t = new int[n + 1][k + 1];

        for (int i = 0; i <= n; i++) {
            t[i][0] = 1;
        }

        for (int i = 1; i <= n; i++) {
            long cumSum = 1;
            for (int j = 1; j <= k; j++) {

                cumSum += t[i - 1][j];
                if (j >= i) {
                    cumSum -= t[i - 1][j - i];
                }
                t[i][j] = (int) (cumSum % M);
            }
        }

        return t[n][k];
    }
}

File: LeetCode/0629-k-inverse-pairs-array/0629-k-inverse-pairs-array.py
class Solution:
    M = 10**9 + 7

    def kInversePairs(self, n: int, k: int) -> int:
        t = [[0] * (k + 1) for _ in range(n + 1)]
        
        for i in range(n + 1):
            t[i][0] = 1

        for i in range(1, n + 1):
            cum_sum = 1
            for j in range(1, k + 1):

                cum_sum += t[i - 1][j]
                if j >= i:
                    cum_sum -= t[i - 1][j - i]

                t[i][j] = cum_sum % self.M

        return t[n][k]